FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt

```diff
@@ -764,117 +800,207 @@ object PgpMsg {
+        val openPgpMetadata = when (block) {
+          is DecryptedAndOrSignedContentMsgBlock -> {
+            block.openPgpMetadata
+          }
 
-      if (block.type in MsgBlock.Type.DECRYPTED_CONTENT_BLOCK_TYPES) {
-        isReplyEncrypted = true
-      }
+          is SignedMsgBlock -> {
+            block.openPgpMetadata
+          }
 
-      when {
-        block.type == MsgBlock.Type.DECRYPTED_ATT -> {
-          resultBlocks.add(block)
+          else -> null
         }
 
-        block.type == MsgBlock.Type.PUBLIC_KEY -> {
-          block.content?.let { source ->
-            try {
-              val keyRings = PgpKey.parseAndNormalizeKeyRings(source)
-              if (keyRings.isNotEmpty()) {
-                resultBlocks.addAll(keyRings.map {
-                  MsgBlockFactory.fromContent(MsgBlock.Type.PUBLIC_KEY, it.armor(null))
-                })
-              } else {
-                resultBlocks.add(
-                  PublicKeyMsgBlock(
-                    content = block.content,
-                    complete = false,
-                    keyDetails = null,
-                    error = MsgBlockError("empty KeyRing")
-                  )
-                )
-              }
-            } catch (ex: Exception) {
-              ex.printStackTrace()
-              resultBlocks.add(
-                PublicKeyMsgBlock(
-                  content = block.content,
-                  complete = false,
-                  keyDetails = null,
-                  error = MsgBlockError(ex.javaClass.simpleName + ": " + ex.message)
-                )
-              )
+        if (!isEncrypted) {
+          isEncrypted = openPgpMetadata?.isEncrypted ?: false
+        }
+
+        if (openPgpMetadata?.isSigned == true) {
+          signedBlockCount++
+
+          if (openPgpMetadata.invalidInbandSignatures.isNotEmpty()
+            || openPgpMetadata.invalidDetachedSignatures.isNotEmpty()
+          ) {
+            val invalidSignatureFailures = openPgpMetadata.invalidInbandSignatures +
+                openPgpMetadata.invalidDetachedSignatures
+
+            hasBadSignatures = invalidSignatureFailures.any {
+              it.validationException.underlyingException != null
+            }
+
+            hasUnverifiedSignatures = invalidSignatureFailures.any {
+              it.validationException.message?.startsWith(
+                "Missing verification certificate", true
+              ) ?: false
             }
           }
-        }
 
-        block.type.isContentBlockType() || MimeUtils.isPlainImgAtt(block) -> {
-          if (block.error != null) {
-            resultBlocks.add(block)
+          if (verifiedSignatures.isEmpty()) {
+            verifiedSignatures.putAll(openPgpMetadata.verifiedSignatures)
           } else {
-            contentBlocks.add(block)
+            val bufferedKeysIds = verifiedSignatures.keys.map { it.keyId }
+            val iteratedKeysIds = openPgpMetadata.verifiedSignatures.keys.map { it.keyId }
```

Comment:
The naming here is not clear - what does buffered and iterated mean? More likely it would be something like `keyIdsOfThisSignature` and `keyIdsOfAllSignatures` or something like that?